build/xtask/src/config.rs

```diff
@@ -4,12 +4,126 @@
+
+        // We use the memory regions from the A and B images to
+        // generate three TOMLS
+
+        let a_file = cfg.parent().unwrap().join(&toml.a_memory);
+        let a = std::fs::read(&a_file)?;
+        let a_toml: MemoryFrag = toml::from_slice(&a)?;
+
+        let b_file = cfg.parent().unwrap().join(&toml.b_memory);
+        let b = std::fs::read(&b_file)?;
+        let b_toml: MemoryFrag = toml::from_slice(&b)?;
+
+        let meta_dir =
+            Path::new("target").join(format!("{}-meta", &toml.base_name));
+
+        std::fs::create_dir_all(&meta_dir)?;
+
+        // Step 1: Bootloader TOML
+        let bootloader_toml = cfg.parent().unwrap().join(&toml.bootloader_toml);
+        let bootloader_bytes = std::fs::read(&bootloader_toml)?;
+
+        let bootloader_path = meta_dir.join("bootloader.toml");
+
+        let mut bootloader = std::fs::File::create(&bootloader_path)?;
+
+        writeln!(bootloader, "# THIS IS AN AUTO GENERATED FILE")?;
+        writeln!(bootloader, "# CHANGE THE TEMPLATE INSTEAD OF THIS FILE")?;
+        writeln!(bootloader, "name = \"{}-bootloader\"", &toml.base_name)?;
+
+        bootloader.write(&bootloader_bytes)?;
+
+        writeln!(bootloader, "[images.a]")?;
+        writeln!(
+            bootloader,
+            "{}",
+            toml::to_string(a_toml.outputs.get("flash").unwrap())?
+        )?;
+
+        writeln!(bootloader, "[images.b]")?;
+        writeln!(
+            bootloader,
+            "{}",
+            toml::to_string(b_toml.outputs.get("flash").unwrap())?
+        )?;
+
+        drop(bootloader);
+
+        // A Image
+        let image_toml = cfg.parent().unwrap().join(&toml.image_toml);
+        let image_toml_bytes = std::fs::read(&image_toml)?;
+        //std::fs::copy(&image_toml, "target/image-a.toml")?;
+        //std::fs::copy(&image_toml, "target/image-b.toml")?;
+
+        let image_a_path = meta_dir.join("image_a.toml");
+        let image_b_path = meta_dir.join("image_b.toml");
+
+        let mut image_a = std::fs::File::create(&image_a_path)?;
+        let mut image_b = std::fs::File::create(&image_b_path)?;
+
+        writeln!(image_a, "# THIS IS AN AUTO GENERATED FILE")?;
+        writeln!(image_a, "# CHANGE THE TEMPLATE INSTEAD OF THIS FILE")?;
+        writeln!(image_a, "name = \"{}-a\"", &toml.base_name)?;
+        image_a.write(&image_toml_bytes)?;
+
+        writeln!(image_b, "# THIS IS AN AUTO GENERATED FILE")?;
+        writeln!(image_b, "# CHANGE THE TEMPLATE INSTEAD OF THIS FILE")?;
+        writeln!(image_b, "name = \"{}-b\"", &toml.base_name)?;
+        image_b.write(&image_toml_bytes)?;
+
+        for (name, e) in a_toml.outputs {
+            writeln!(image_a, "[outputs.{}]", name)?;
+            writeln!(image_a, "{}", toml::to_string(&e)?)?;
+        }
+
+        for (name, e) in b_toml.outputs {
+            writeln!(image_b, "[outputs.{}]", name)?;
+            writeln!(image_b, "{}", toml::to_string(&e)?)?;
+        }
```

Comment:
Maybe move duplicate code into a closure here, e.g.
```suggestion
        let write_image_toml = |suffix, mem: MemoryFrag| -> Result<PathBuf> {
            let image_path = meta_dir.join(&format!("image_{}.toml", suffix));
            let mut image = std::fs::File::create(&image_path)?;

            writeln!(image, "# THIS IS AN AUTO GENERATED FILE")?;
            writeln!(image, "# CHANGE THE TEMPLATE INSTEAD OF THIS FILE")?;
            writeln!(image, "name = \"{}-{}\"", &toml.base_name, suffix)?;
            image.write(&image_toml_bytes)?;

            for (name, e) in mem.outputs {
                writeln!(image, "[outputs.{}]", name)?;
                writeln!(image, "{}", toml::to_string(&e)?)?;
            }
            Ok(image_path)
        };
        let image_a_path = write_image_toml('a', a_toml)?;
        let image_b_path = write_image_toml('b', b_toml)?;
```